{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "1108d367-a447-4965-9c1c-aee2fde4a210",
   "metadata": {},
   "source": [
    "# Pulling data from MGnify\n",
    "\n",
    "Drew Honson\n",
    "\n",
    "13 October 2023"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "18f4ffce-ed3f-411f-a90e-e77b163da153",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "from tqdm import tqdm\n",
    "import os, re, sys, wget\n",
    "\n",
    "def split_taxa(x):\n",
    "    tax_dict = {'k':'Kingdom',\n",
    "                'p':'Phylum',\n",
    "                'c':'Class',\n",
    "                'o':'Order',\n",
    "                'f':'Family',\n",
    "                'g':'Genus',\n",
    "                's':'Species'}\n",
    "    \n",
    "    tax_ls = x.split(';')[1:]\n",
    "    tax_ls = [i.split('__') for i in tax_ls]\n",
    "    \n",
    "    out_dict = {}\n",
    "\n",
    "    for i,v in enumerate(tax_ls):\n",
    "        #v[0] = tax_dict[v[0]]\n",
    "        out_dict[tax_dict[v[0]]] = v[1]\n",
    "    \n",
    "    return out_dict"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1bafa6ff-64bc-4f92-9dfc-b7668ebf96c3",
   "metadata": {},
   "source": [
    "## Use wget to pull data\n",
    "\n",
    "This study examined the diversity of mouse microbiomes at eight sites in Western Europe. The study can be found [here.](https://www.ebi.ac.uk/metagenomics/studies/MGYS00000516#overview)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "f6881df3-987c-46a1-a95f-e7bf35798986",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-1 / unknown...........................................] 156156 / 156156"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#SampleID</th>\n",
       "      <th>ERR197719</th>\n",
       "      <th>ERR197720</th>\n",
       "      <th>ERR197721</th>\n",
       "      <th>ERR197722</th>\n",
       "      <th>ERR197723</th>\n",
       "      <th>ERR197724</th>\n",
       "      <th>ERR197725</th>\n",
       "      <th>ERR197726</th>\n",
       "      <th>ERR197727</th>\n",
       "      <th>...</th>\n",
       "      <th>ERR197910</th>\n",
       "      <th>ERR197911</th>\n",
       "      <th>ERR197912</th>\n",
       "      <th>ERR197913</th>\n",
       "      <th>ERR197914</th>\n",
       "      <th>ERR197915</th>\n",
       "      <th>ERR197916</th>\n",
       "      <th>ERR197917</th>\n",
       "      <th>ERR197918</th>\n",
       "      <th>ERR197919</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Root;k__Bacteria;p__Acidobacteria;c__Acidobact...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Root;k__Bacteria;p__Acidobacteria;c__Acidobact...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Root;k__Bacteria;p__Actinobacteria;c__Actinoba...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Root;k__Bacteria;p__Actinobacteria;c__Actinoba...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Root;k__Bacteria;p__Actinobacteria;c__Actinoba...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 202 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                           #SampleID  ERR197719  ERR197720  \\\n",
       "0  Root;k__Bacteria;p__Acidobacteria;c__Acidobact...          0          0   \n",
       "1  Root;k__Bacteria;p__Acidobacteria;c__Acidobact...          0          0   \n",
       "2  Root;k__Bacteria;p__Actinobacteria;c__Actinoba...          0          0   \n",
       "3  Root;k__Bacteria;p__Actinobacteria;c__Actinoba...          0          0   \n",
       "4  Root;k__Bacteria;p__Actinobacteria;c__Actinoba...          0          0   \n",
       "\n",
       "   ERR197721  ERR197722  ERR197723  ERR197724  ERR197725  ERR197726  \\\n",
       "0          0          0          0          0          0          0   \n",
       "1          0          0          0          0          0          0   \n",
       "2          0          0          0          0          0          0   \n",
       "3          0          0          0          0          0          0   \n",
       "4          0          0          0          0          0          0   \n",
       "\n",
       "   ERR197727  ...  ERR197910  ERR197911  ERR197912  ERR197913  ERR197914  \\\n",
       "0          0  ...          0          0          0          0          0   \n",
       "1          0  ...          0          0          0          0          0   \n",
       "2          0  ...          0          0          0          0          0   \n",
       "3          0  ...          0          0          0          0          0   \n",
       "4          0  ...          0          0          0          0          0   \n",
       "\n",
       "   ERR197915  ERR197916  ERR197917  ERR197918  ERR197919  \n",
       "0          0          0          0          0          0  \n",
       "1          0          0          0          0          0  \n",
       "2          0          0          0          0          0  \n",
       "3          0          0          0          0          0  \n",
       "4          0          0          0          0          0  \n",
       "\n",
       "[5 rows x 202 columns]"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Make data directory\n",
    "DATA_DIR = 'data'\n",
    "\n",
    "if os.path.exists(DATA_DIR):\n",
    "    pass\n",
    "else:\n",
    "    os.mkdir(DATA_DIR)\n",
    "    \n",
    "# Get taxonomy, sample, and run information\n",
    "TAXON_URL = \"https://www.ebi.ac.uk/metagenomics/api/v1/studies/MGYS00000516/pipelines/2.0/file/ERP001970_taxonomy_abundances_v2.0.tsv\"\n",
    "SAMPLE_URL = \"https://www.ebi.ac.uk/metagenomics/api/v1/studies/MGYS00000516/analyses?include=sample&ordering=&format=csv\"\n",
    "\n",
    "TAXON_PATH = wget.download(TAXON_URL, out = DATA_DIR)\n",
    "SAMPLE_PATH = wget.download(SAMPLE_URL, out = DATA_DIR)\n",
    "\n",
    "# Read taxonomy data\n",
    "TAXON_DATA = pd.read_csv(TAXON_PATH, sep='\\t')\n",
    "\n",
    "# Drop root\n",
    "TAXON_DATA.drop(0,inplace=True)\n",
    "TAXON_DATA.reset_index(drop=True,inplace=True)\n",
    "TAXON_DATA.to_csv(TAXON_PATH, index=False)\n",
    "\n",
    "TAXON_DATA.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "6fe2f11a-c13d-4b87-a9de-b878e3c457ec",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#SampleID</th>\n",
       "      <th>ERR197719</th>\n",
       "      <th>ERR197720</th>\n",
       "      <th>ERR197721</th>\n",
       "      <th>ERR197722</th>\n",
       "      <th>ERR197723</th>\n",
       "      <th>ERR197724</th>\n",
       "      <th>ERR197725</th>\n",
       "      <th>ERR197726</th>\n",
       "      <th>ERR197727</th>\n",
       "      <th>...</th>\n",
       "      <th>ERR197917</th>\n",
       "      <th>ERR197918</th>\n",
       "      <th>ERR197919</th>\n",
       "      <th>Kingdom</th>\n",
       "      <th>Phylum</th>\n",
       "      <th>Class</th>\n",
       "      <th>Order</th>\n",
       "      <th>Family</th>\n",
       "      <th>Genus</th>\n",
       "      <th>Species</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>288</th>\n",
       "      <td>Root;k__Bacteria;p__TM7;c__TM7-1;o__;f__;g__;s__</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>TM7</td>\n",
       "      <td>TM7-1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>289</th>\n",
       "      <td>Root;k__Bacteria;p__Tenericutes;c__Mollicutes;...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>18</td>\n",
       "      <td>12</td>\n",
       "      <td>27</td>\n",
       "      <td>10</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Tenericutes</td>\n",
       "      <td>Mollicutes</td>\n",
       "      <td>Anaeroplasmatales</td>\n",
       "      <td>Anaeroplasmataceae</td>\n",
       "      <td>Anaeroplasma</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>290</th>\n",
       "      <td>Root;k__Bacteria;p__Tenericutes;c__Mollicutes;...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Tenericutes</td>\n",
       "      <td>Mollicutes</td>\n",
       "      <td>Mycoplasmatales</td>\n",
       "      <td>Mycoplasmataceae</td>\n",
       "      <td>Mycoplasma</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>291</th>\n",
       "      <td>Root;k__Bacteria;p__Tenericutes;c__Mollicutes;...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Tenericutes</td>\n",
       "      <td>Mollicutes</td>\n",
       "      <td>RF39</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>292</th>\n",
       "      <td>Root;k__Bacteria;p__Verrucomicrobia;c__Verruco...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Verrucomicrobia</td>\n",
       "      <td>Verrucomicrobiae</td>\n",
       "      <td>Verrucomicrobiales</td>\n",
       "      <td>Verrucomicrobiaceae</td>\n",
       "      <td>Akkermansia</td>\n",
       "      <td>muciniphila</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 209 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                             #SampleID  ERR197719  ERR197720  \\\n",
       "288   Root;k__Bacteria;p__TM7;c__TM7-1;o__;f__;g__;s__          0          0   \n",
       "289  Root;k__Bacteria;p__Tenericutes;c__Mollicutes;...          0          0   \n",
       "290  Root;k__Bacteria;p__Tenericutes;c__Mollicutes;...          0          0   \n",
       "291  Root;k__Bacteria;p__Tenericutes;c__Mollicutes;...          0          0   \n",
       "292  Root;k__Bacteria;p__Verrucomicrobia;c__Verruco...          0          0   \n",
       "\n",
       "     ERR197721  ERR197722  ERR197723  ERR197724  ERR197725  ERR197726  \\\n",
       "288          0          0          0          0          0          0   \n",
       "289          0          2          0         18         12         27   \n",
       "290          0          0          0          0          0          0   \n",
       "291          1          0          0          0          0          0   \n",
       "292          0          0          0          0          0          0   \n",
       "\n",
       "     ERR197727  ...  ERR197917  ERR197918  ERR197919   Kingdom  \\\n",
       "288          0  ...          0          0          0  Bacteria   \n",
       "289         10  ...          0          0          0  Bacteria   \n",
       "290          0  ...          0          0          0  Bacteria   \n",
       "291          0  ...          0          2          0  Bacteria   \n",
       "292          0  ...          0          0          0  Bacteria   \n",
       "\n",
       "              Phylum             Class               Order  \\\n",
       "288              TM7             TM7-1                       \n",
       "289      Tenericutes        Mollicutes   Anaeroplasmatales   \n",
       "290      Tenericutes        Mollicutes     Mycoplasmatales   \n",
       "291      Tenericutes        Mollicutes                RF39   \n",
       "292  Verrucomicrobia  Verrucomicrobiae  Verrucomicrobiales   \n",
       "\n",
       "                  Family         Genus      Species  \n",
       "288                                                  \n",
       "289   Anaeroplasmataceae  Anaeroplasma               \n",
       "290     Mycoplasmataceae    Mycoplasma               \n",
       "291                                                  \n",
       "292  Verrucomicrobiaceae   Akkermansia  muciniphila  \n",
       "\n",
       "[5 rows x 209 columns]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Export reads with taxa\n",
    "for i in TAXON_DATA.index:\n",
    "    tdict = split_taxa(TAXON_DATA.loc[i,'#SampleID'])\n",
    "    \n",
    "    for t in tdict:\n",
    "        TAXON_DATA.loc[i,t] = tdict[t]\n",
    "        \n",
    "TAXON_DATA.tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "73c1deb7-562a-4df7-ba31-cb9f9f5e1d3f",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Export \n",
    "TAXON_DATA.to_csv(DATA_DIR+'/sampleWithTaxonomy.csv',index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "722bb514-2a1d-46ee-b196-df40d04969c5",
   "metadata": {},
   "source": [
    "## Make a convenient dataframe for exploring taxa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "76039432-a815-4afe-a931-8bf86917ee22",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Kingdom</th>\n",
       "      <th>Phylum</th>\n",
       "      <th>Class</th>\n",
       "      <th>Order</th>\n",
       "      <th>Family</th>\n",
       "      <th>Genus</th>\n",
       "      <th>Species</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Acidobacteria</td>\n",
       "      <td>Acidobacteria-6</td>\n",
       "      <td>iii1-15</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Acidobacteria</td>\n",
       "      <td>Acidobacteria-6</td>\n",
       "      <td>iii1-15</td>\n",
       "      <td>mb2424</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>Actinomycetales</td>\n",
       "      <td>ACK-M1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>Actinomycetales</td>\n",
       "      <td>Bogoriellaceae</td>\n",
       "      <td>Georgenia</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>Actinomycetales</td>\n",
       "      <td>Brevibacteriaceae</td>\n",
       "      <td>Brevibacterium</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Kingdom          Phylum            Class            Order  \\\n",
       "0  Bacteria   Acidobacteria  Acidobacteria-6          iii1-15   \n",
       "1  Bacteria   Acidobacteria  Acidobacteria-6          iii1-15   \n",
       "2  Bacteria  Actinobacteria   Actinobacteria  Actinomycetales   \n",
       "3  Bacteria  Actinobacteria   Actinobacteria  Actinomycetales   \n",
       "4  Bacteria  Actinobacteria   Actinobacteria  Actinomycetales   \n",
       "\n",
       "              Family           Genus Species  \n",
       "0                                             \n",
       "1             mb2424                          \n",
       "2             ACK-M1                          \n",
       "3     Bogoriellaceae       Georgenia          \n",
       "4  Brevibacteriaceae  Brevibacterium          "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Taxonomy dataframe\n",
    "taxons = [i.split(';')[1:] for i in TAXON_DATA.loc[1:,'#SampleID']]\n",
    "\n",
    "for i,v in enumerate(taxons):\n",
    "    taxons[i] = [i.split('__')[-1] for i in v]\n",
    "\n",
    "taxonomy = pd.DataFrame(taxons, \n",
    "                       columns=['Kingdom','Phylum','Class',\n",
    "                                'Order','Family','Genus','Species'])\n",
    "\n",
    "taxonomy.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "74368086-3e1b-4ba4-887f-ef2eaa963741",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Export \n",
    "taxonomy.to_csv('data/taxonomy.csv',index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1bb45486-7661-4980-ad2e-2bc62b2a44f8",
   "metadata": {},
   "source": [
    "## Pull individual sample data\n",
    "\n",
    "This study has 200 samples total. The following code extracts the location data for each of these. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "99a09110-cd4a-40ff-8d68-bbd56b85fbe3",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|â–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆ| 201/201 [05:53<00:00,  1.76s/it]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>run</th>\n",
       "      <th>sample</th>\n",
       "      <th>sampfile</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ERR197740</td>\n",
       "      <td>https://www.ebi.ac.uk/metagenomics/api/v1/samp...</td>\n",
       "      <td>data/ERS194057</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ERR197741</td>\n",
       "      <td>https://www.ebi.ac.uk/metagenomics/api/v1/samp...</td>\n",
       "      <td>data/ERS194058</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ERR197742</td>\n",
       "      <td>https://www.ebi.ac.uk/metagenomics/api/v1/samp...</td>\n",
       "      <td>data/ERS194059</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ERR197743</td>\n",
       "      <td>https://www.ebi.ac.uk/metagenomics/api/v1/samp...</td>\n",
       "      <td>data/ERS194060</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ERR197744</td>\n",
       "      <td>https://www.ebi.ac.uk/metagenomics/api/v1/samp...</td>\n",
       "      <td>data/ERS194061</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         run                                             sample  \\\n",
       "0  ERR197740  https://www.ebi.ac.uk/metagenomics/api/v1/samp...   \n",
       "1  ERR197741  https://www.ebi.ac.uk/metagenomics/api/v1/samp...   \n",
       "2  ERR197742  https://www.ebi.ac.uk/metagenomics/api/v1/samp...   \n",
       "3  ERR197743  https://www.ebi.ac.uk/metagenomics/api/v1/samp...   \n",
       "4  ERR197744  https://www.ebi.ac.uk/metagenomics/api/v1/samp...   \n",
       "\n",
       "         sampfile  \n",
       "0  data/ERS194057  \n",
       "1  data/ERS194058  \n",
       "2  data/ERS194059  \n",
       "3  data/ERS194060  \n",
       "4  data/ERS194061  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Download sample data\n",
    "SAMPLE_DATA = pd.read_csv(SAMPLE_PATH, usecols=['run','sample'])\n",
    "\n",
    "SAMPLE_DATA['run'] = [i.split('/')[-1] for i in SAMPLE_DATA['run']]\n",
    "SAMPLE_DATA['run'] = [i.split('?')[0] for i in SAMPLE_DATA['run']]\n",
    "\n",
    "for s in tqdm(SAMPLE_DATA.index):\n",
    "    SAMPLE_DATA.loc[s,'sampfile'] = wget.download(SAMPLE_DATA.loc[s,'sample'], out = DATA_DIR, bar=False)\n",
    "    \n",
    "SAMPLE_DATA.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "0d6db8a0-61c0-4944-9c66-768138396632",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|â–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆ| 201/201 [00:00<00:00, 236.52it/s]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>run</th>\n",
       "      <th>sample</th>\n",
       "      <th>sampfile</th>\n",
       "      <th>geo_loc_name</th>\n",
       "      <th>latitude</th>\n",
       "      <th>longitude</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ERR197740</td>\n",
       "      <td>https://www.ebi.ac.uk/metagenomics/api/v1/samp...</td>\n",
       "      <td>data/ERS194057</td>\n",
       "      <td>France</td>\n",
       "      <td>47.4532</td>\n",
       "      <td>0.5949</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ERR197741</td>\n",
       "      <td>https://www.ebi.ac.uk/metagenomics/api/v1/samp...</td>\n",
       "      <td>data/ERS194058</td>\n",
       "      <td>France</td>\n",
       "      <td>47.4532</td>\n",
       "      <td>0.5949</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ERR197742</td>\n",
       "      <td>https://www.ebi.ac.uk/metagenomics/api/v1/samp...</td>\n",
       "      <td>data/ERS194059</td>\n",
       "      <td>France</td>\n",
       "      <td>47.4532</td>\n",
       "      <td>0.5949</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ERR197743</td>\n",
       "      <td>https://www.ebi.ac.uk/metagenomics/api/v1/samp...</td>\n",
       "      <td>data/ERS194060</td>\n",
       "      <td>France</td>\n",
       "      <td>47.4532</td>\n",
       "      <td>0.5949</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ERR197744</td>\n",
       "      <td>https://www.ebi.ac.uk/metagenomics/api/v1/samp...</td>\n",
       "      <td>data/ERS194061</td>\n",
       "      <td>France</td>\n",
       "      <td>47.4532</td>\n",
       "      <td>0.5949</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         run                                             sample  \\\n",
       "0  ERR197740  https://www.ebi.ac.uk/metagenomics/api/v1/samp...   \n",
       "1  ERR197741  https://www.ebi.ac.uk/metagenomics/api/v1/samp...   \n",
       "2  ERR197742  https://www.ebi.ac.uk/metagenomics/api/v1/samp...   \n",
       "3  ERR197743  https://www.ebi.ac.uk/metagenomics/api/v1/samp...   \n",
       "4  ERR197744  https://www.ebi.ac.uk/metagenomics/api/v1/samp...   \n",
       "\n",
       "         sampfile geo_loc_name  latitude  longitude  \n",
       "0  data/ERS194057       France   47.4532     0.5949  \n",
       "1  data/ERS194058       France   47.4532     0.5949  \n",
       "2  data/ERS194059       France   47.4532     0.5949  \n",
       "3  data/ERS194060       France   47.4532     0.5949  \n",
       "4  data/ERS194061       France   47.4532     0.5949  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "for s in tqdm(SAMPLE_DATA.index):\n",
    "    df = pd.read_csv(SAMPLE_DATA.loc[s,'sampfile'])\n",
    "    SAMPLE_DATA.loc[s,'geo_loc_name'] = df['geo_loc_name'][0]\n",
    "    SAMPLE_DATA.loc[s,'latitude'] = df['latitude'][0]\n",
    "    SAMPLE_DATA.loc[s,'longitude'] = df['longitude'][0]\n",
    "    \n",
    "SAMPLE_DATA.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "37967bd1-2460-4f18-89c1-3346db97bf58",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Simplify dataframe\n",
    "SAMPLE_DATA.drop('sample',axis=1,inplace=True)\n",
    "SAMPLE_DATA['sample'] = [i.split('/')[-1] for i in SAMPLE_DATA['sampfile']]\n",
    "\n",
    "# Clean up location files\n",
    "for i in SAMPLE_DATA['sampfile']:\n",
    "    os.remove(i)\n",
    "SAMPLE_DATA.drop('sampfile',axis=1,inplace=True)\n",
    "\n",
    "# Export data\n",
    "SAMPLE_DATA.to_csv('data/sampleLocs.csv',index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "3be44083-9dae-4096-a034-7a0ebf874cac",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>run</th>\n",
       "      <th>geo_loc_name</th>\n",
       "      <th>latitude</th>\n",
       "      <th>longitude</th>\n",
       "      <th>sample</th>\n",
       "      <th>coord</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ERR197740</td>\n",
       "      <td>France</td>\n",
       "      <td>47.4532</td>\n",
       "      <td>0.5949</td>\n",
       "      <td>ERS194057</td>\n",
       "      <td>(47.4532, 0.5949)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ERR197741</td>\n",
       "      <td>France</td>\n",
       "      <td>47.4532</td>\n",
       "      <td>0.5949</td>\n",
       "      <td>ERS194058</td>\n",
       "      <td>(47.4532, 0.5949)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ERR197742</td>\n",
       "      <td>France</td>\n",
       "      <td>47.4532</td>\n",
       "      <td>0.5949</td>\n",
       "      <td>ERS194059</td>\n",
       "      <td>(47.4532, 0.5949)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ERR197743</td>\n",
       "      <td>France</td>\n",
       "      <td>47.4532</td>\n",
       "      <td>0.5949</td>\n",
       "      <td>ERS194060</td>\n",
       "      <td>(47.4532, 0.5949)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ERR197744</td>\n",
       "      <td>France</td>\n",
       "      <td>47.4532</td>\n",
       "      <td>0.5949</td>\n",
       "      <td>ERS194061</td>\n",
       "      <td>(47.4532, 0.5949)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         run geo_loc_name  latitude  longitude     sample              coord\n",
       "0  ERR197740       France   47.4532     0.5949  ERS194057  (47.4532, 0.5949)\n",
       "1  ERR197741       France   47.4532     0.5949  ERS194058  (47.4532, 0.5949)\n",
       "2  ERR197742       France   47.4532     0.5949  ERS194059  (47.4532, 0.5949)\n",
       "3  ERR197743       France   47.4532     0.5949  ERS194060  (47.4532, 0.5949)\n",
       "4  ERR197744       France   47.4532     0.5949  ERS194061  (47.4532, 0.5949)"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Update location data\n",
    "locations = pd.read_csv('data/sampleLocs.csv')\n",
    "\n",
    "# Add coordinates\n",
    "loc_ls = list(zip(locations['latitude'],locations['longitude']))\n",
    "loc_ls = [str(i) for i in loc_ls]\n",
    "locations['coord'] = loc_ls\n",
    "\n",
    "locations.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "140d693e-69c1-4e62-83d1-981cf82dae82",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Export locations\n",
    "locations.to_csv('data/sampleLocs.csv',index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "16d05297-d8df-4fb4-8bcc-f531c1113ce5",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Python implementation: CPython\n",
      "Python version       : 3.11.5\n",
      "IPython version      : 8.15.0\n",
      "\n",
      "wget  : 3.2\n",
      "sys   : 3.11.5 (main, Sep 11 2023, 08:19:27) [Clang 14.0.6 ]\n",
      "pandas: 2.0.3\n",
      "re    : 2.2.1\n",
      "\n"
     ]
    }
   ],
   "source": [
    "%load_ext watermark\n",
    "%watermark -v --iversions"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
